### PR TITLE
Added options to GAP functions as keyword arguments

### PR DESCRIPTION
**Pull request by sebasguts**
It would be cool to have an example. Can you point me to a GAP function that does something with options?

### CODE REVIEW

**Comment by ThomasBreuer** [comment]
Here is an example where options change the output.

```
gap> x:= Indeterminate( Rationals );;
gap> f:= x^4 + 1;;
gap> IdealDecompositionsOfPolynomial( f );
[ [ x_1^2-2, x_1^3-x_1 ], [ x_1^2+1, x_1^2 ], [ x_1^2+2, x_1^3+x_1 ] ]
gap> IdealDecompositionsOfPolynomial( f : onlyone );
[ [ x_1^2-2, x_1^3-x_1 ] ]
```

**Comment by sebasguts** [comment]
Thank you very much. This works.

File: test/basics.jl [hunk 1]
```diff
@@ -50,4 +50,10 @@ end
 
     @test GAP.Globals.IdFunc(2^62) isa Int64
 
+    x = GAP.Globals.Indeterminate( GAP.Globals.Rationals )
+    f = x^4+1
+    i1 = GAP.Globals.IdealDecompositionsOfPolynomial( f )
+    i2 = GAP.Globals.IdealDecompositionsOfPolynomial( f , onlyone = true)
```
Comment (fingolfin) [inline test/basics.jl]:
```suggestion
    i2 = GAP.Globals.IdealDecompositionsOfPolynomial( f , onlyone = true )
```
... makes this change locally consistent. But of course it is inconsistent with the surrounding code, which does not put spaces around parenthesis like that...

I wonder whether there is a usable code formatter for Julia? Last time I looked, I couldn't find one.

File: pkg/GAPJulia/JuliaInterface/julia/ccalls.jl [hunk 2]
```diff
@@ -125,10 +125,22 @@ end
...
+    options = false
+    if length(kwargs) > 0
+        kwargs_dict = Dict(kwargs)
+        kwargs_rec = julia_to_gap(kwargs_dict)
+        Globals.PushOptions(kwargs_rec)
+        options = true
+    end
+    result = ccall(:call_gap_func, Any, (MPtr, Any), func, args)
+    if options
+        Globals.PopOptions()
```
Comment (fingolfin) [inline pkg/GAPJulia/JuliaInterface/julia/ccalls.jl]: Perhaps there should also be a test involving options which runs into a GAP error, to see if the options stack is correctly cleared in this case, too?

File: test/basics.jl (hunk 1 above)
Comment (fingolfin) [reply test/basics.jl]: Perhaps add a check that `GAP.Globals.OptionsStack` is empty at this point?

Comment (sebasguts) [reply test/basics.jl]: Will add that.
I checked for a formatter, and was not able to find one, unfortunately . . .

File: pkg/GAPJulia/JuliaInterface/julia/ccalls.jl (hunk 2 above)
Comment (sebasguts) [reply pkg/GAPJulia/JuliaInterface/julia/ccalls.jl]: I added one, and of course discovered that the stack was not reset. Now, I reset the stack.

File: pkg/GAPJulia/JuliaInterface/julia/ccalls.jl [hunk 3]
```diff
@@ -125,10 +125,30 @@ end
 > There is no argument number checking here, all checks on the arguments
 > are done by GAP itself.
 """
-(func::MPtr)(args...) = call_gap_func(func,args...)
-
-function call_gap_func(func::MPtr,args...)
-    return ccall(:call_gap_func, Any, (MPtr, Any), func, args)
+(func::MPtr)(args...; kwargs...) = call_gap_func(func,args...; kwargs...)
```
Comment (fingolfin) [inline pkg/GAPJulia/JuliaInterface/julia/ccalls.jl]:
```suggestion
(func::MPtr)(args...; kwargs...) = call_gap_func(func, args...; kwargs...)
```

File: pkg/GAPJulia/JuliaInterface/julia/ccalls.jl [hunk 4]
```diff
@@ -125,10 +125,30 @@ end
...
+
+function call_gap_func(func::MPtr,args...; kwargs...)
```
Comment (fingolfin) [inline pkg/GAPJulia/JuliaInterface/julia/ccalls.jl]:
```suggestion
function call_gap_func(func::MPtr, args...; kwargs...)
```

**Comment by sebasguts** [comment]
Thx. I applied them :)

**Pull request merged**